state-commitments/optimized/history_commitment.go

```diff
@@ -24,17 +24,46 @@ func NewCommitment(leaves []common.Hash, virtual uint64) (*Commitment, error) {
 	if len(leaves) == 0 {
 		return nil, errors.New("must commit to at least one leaf")
 	}
+	if virtual == 0 {
+		return nil, errors.New("virtual size cannot be zero")
+	}
 	comm := NewCommitter()
-	root, err := comm.ComputeRoot(leaves, virtual)
+	leavesForProofs := make([]common.Hash, len(leaves))
+	leavesForCommit := make([]common.Hash, len(leaves))
+	copy(leavesForProofs, leaves)
+	firstLeaf := leavesForProofs[0]
+	lastLeaf := leavesForProofs[len(leavesForProofs)-1]
+	for i := 0; i < len(leavesForProofs); i++ {
+		result, err := comm.hash(leavesForProofs[i][:])
+		if err != nil {
+			return nil, err
+		}
+		leavesForCommit[i] = result
+		leavesForProofs[i] = result
+	}
+	var firstLeafProof, lastLeafProof []common.Hash
+	var err error
+	if isPowTwo(virtual) {
```

Comment:
Do you want to return an error if it's not a power of two?. Also there's no check against `virtual >= len(leaves)` in this function. Currently this relies in the called functions that actually check for this. But it's better to add this cheap check early here. Also,  here, if it's not a power of two, higher than the length of the leaves, it may be undefined behavior downstream.

Reply:
The problem is that you are calling then the compute root function with a virtual that is not a power of two and this is undefined. I think you want to explciitly handle the non-power of two case then if you don't want to throw an error

Reply:
Im not sure I follow. The compute root function can take in any virtual that is >= len(leaves), so I could call it with virtual of 6, 7, or 13, etc. That is a requirement for our history commits